narwhals/_pandas_like/dataframe.py

```diff
@@ -167,16 +167,47 @@ def filter(
 
     def with_columns(
         self,
-        *exprs: IntoPandasExpr | Iterable[IntoPandasExpr],
+        *exprs: IntoPandasExpr,
         **named_exprs: IntoPandasExpr,
     ) -> Self:
+        from narwhals._pandas_like.expr import PandasExpr
+
+        index = self._dataframe.index
         new_series = evaluate_into_exprs(self, *exprs, **named_exprs)
-        df = self._dataframe.assign(
-            **{
-                series.name: validate_dataframe_comparand(self._dataframe.index, series)
-                for series in new_series
-            }
+        # If the inputs are all Expressions which return full columns
+        # (as opposed to scalars), we can use a fast path (concat, instead of assign).
+        # We can't use the fastpath if any input is not an expression (e.g.
+        # if it's a Series) because then we might be changing its flags.
+        # See `test_memmap` for an example of where this is necessary.
+        fast_path = (
+            all(s.len() > 1 for s in new_series)
+            and all(isinstance(x, PandasExpr) for x in exprs)
+            and all(isinstance(x, PandasExpr) for (_, x) in named_exprs.items())
         )
+
+        if fast_path:
+            new_names = {s.name: s for s in new_series}
+            to_concat = []
+            # Make sure to preserve column order
+            for s in self._dataframe.columns:
+                if s in new_names:
+                    to_concat.append(
+                        validate_dataframe_comparand(index, new_names.pop(s))
```

Comment:
The "modifying while looping" scares the s**t out of me 🥲

Reply:
😄 I see what you mean, but we're not modifying what we're iterating over, but another variable

we iterate over `self._dataframe.columns`, and modify `new_names`